exonum/src/runtime/types.rs

```diff
@@ -447,13 +482,49 @@ pub struct InstanceState {
     /// Service instance specification.
     pub spec: InstanceSpec,
     /// Service instance activity status.
-    pub status: InstanceStatus,
+    #[protobuf_convert(with = "InstanceStatus")]
+    pub status: Option<InstanceStatus>,
+    /// Pending status of instance if the value is not `None`.
+    #[protobuf_convert(with = "InstanceStatus")]
+    pub pending_status: Option<InstanceStatus>,
 }
 
 impl InstanceState {
     /// Creates a new instance state with the given specification and status.
     pub fn new(spec: InstanceSpec, status: InstanceStatus) -> Self {
-        Self { spec, status }
+        Self {
+            spec,
+            status: Some(status),
+            pending_status: None,
+        }
+    }
+
+    /// Indicates whether the service instance current or pending status is active.
+    pub fn is_active(&self) -> bool {
```

Comment:
I'm not sure that this is a good name; IMO, it sort of implies that the service is active now. I'd suggest to have a method returning status accounting for possible pending status, i.e.,

```rust
self.pending_status.or(self.status)
```

...and then act on this status depending on the caller.